File: 1 - metodosBusca/pacoteBusca_JomiHubner/src/busca/SubidaMontanha.java
package busca;

import java.util.List;

/**
 *   Algoritmos de Busca (geral, qquer problema)
 *   
 *   Busca a solução por "subida da montanha"
 *                        ------------------
 *
 *   @author Jomi Fred Hübner
 */
public class SubidaMontanha extends BuscaHeuristica {
    
    /** busca sem mostrar status */
    public SubidaMontanha() {
    }
    
    /** busca mostrando status */
    public SubidaMontanha(MostraStatusConsole ms) {
        super(ms);
    }

    public Nodo busca(Estado corrente) {
        status.inicia();
        initFechados();

        Estado melhor = corrente;
        while (!parar && corrente != null) {
            
            List<Estado> filhos = corrente.sucessores();
            if (filhos.size() == 0) {
                corrente = ((Aleatorio)corrente).geraAleatorio(); // começa em outro lugar aleatório
                continue;
            }

            // acha o menor h entre os filhos
            Estado melhorFilho = null;
            for (Estado e: filhos) {
                if (melhorFilho == null) {
                    melhorFilho = e;
                } else if ( ((Heuristica)e).h() < ((Heuristica)melhorFilho).h()) {
                    melhorFilho = e;
                }
            }
            
            status.nroVisitados++;
            
            // se não tem filho melhor que corrente
            if (((Heuristica)melhorFilho).h() >= ((Heuristica)corrente).h()) {
                if (corrente.ehMeta()) {
                    status.termina(true);
                    return new Nodo(corrente, null);
                } else { // máximo local
                    if (((Heuristica)corrente).h() < ((Heuristica)melhor).h()) {
                        melhor = corrente;
                    }
                    corrente = ((Aleatorio)corrente).geraAleatorio(); // começa em outro lugar aleatório
                }
            } else { // tem filho melhor que corrente
                corrente = melhorFilho;
            }
        }
        status.termina(false);
        return null;
    }
    
    public String toString() {
    	return "BSM - busca com subida da montanha";
    }
}

File: 1 - metodosBusca/python/app/core/busca/subida_montanha.py
from .busca_heuristica import BuscaHeuristica
from .nodo import Nodo

class SubidaMontanha(BuscaHeuristica):
    def __init__(self, mostra_status_console=None):
        super().__init__(mostra_status_console)

    def busca(self, corrente):
        self.status.inicia()
        self.init_fechados()

        melhor = corrente
        while not self.parar and corrente is not None:
            filhos = corrente.sucessores()
            if not filhos:
                corrente = corrente.gera_aleatorio()  # começa em outro lugar aleatório
                continue

            # acha o menor h entre os filhos
            melhor_filho = None
            for e in filhos:
                if melhor_filho is None or e.h() < melhor_filho.h():
                    melhor_filho = e

            self.status.nro_visitados += 1

            # se não tem filho melhor que corrente
            if melhor_filho.h() >= corrente.h():
                if corrente.eh_meta():
                    self.status.termina(True)
                    return Nodo(corrente, None)
                else:  # máximo local
                    if corrente.h() < melhor.h():
                        melhor = corrente
                    corrente = corrente.gera_aleatorio()  # começa em outro lugar aleatório
            else:  # tem filho melhor que corrente
                corrente = melhor_filho

        self.status.termina(False)
        return None

    def __str__(self):
        return "BSM - busca com subida da montanha"